File: src/main/java/org/gga/graph/search/dfs/VertexEventVisitor.java
package org.gga.graph.search.dfs;

import org.gga.graph.Graph;

/**
 * @author mike
 */
public abstract class VertexEventVisitor extends AbstractDfsVisitor {
    public enum VertexEvent {
        INITIALIZE_VERTEX,
        START_VERTEX,
        DISCOVER_VERTEX,
        FINISH_VERTEX
    }

    public abstract void onEvent(int v, Graph graph, VertexEvent event);

    @Override
    public void initializeVertex(int v, Graph graph) {
        onEvent(v, graph, VertexEvent.INITIALIZE_VERTEX);
    }

    @Override
    public void startVertex(int v, Graph graph) {
        onEvent(v, graph, VertexEvent.START_VERTEX);
    }

    @Override
    public void discoverVertex(int v, Graph graph) {
        onEvent(v, graph, VertexEvent.DISCOVER_VERTEX);
    }

    @Override
    public void finishVertex(int v, Graph graph) {
        onEvent(v, graph, VertexEvent.FINISH_VERTEX);
    }
}

File: src/main/scala/org/gga/graph/search/dfs/VertexEventVisitor.scala
package org.gga.graph.search.dfs

import org.gga.graph.Graph
import org.gga.graph.search.dfs.VertexEventVisitor.VertexEvent

/**
 * @author mike
 */
object VertexEventVisitor {
  object VertexEvent extends Enumeration {
    val INITIALIZE_VERTEX = Value
    val START_VERTEX = Value
    val DISCOVER_VERTEX = Value
    val FINISH_VERTEX = Value
  }
}

trait VertexEventVisitor extends AbstractDfsVisitor {
  def onEvent(v: Int, graph: Graph, event: VertexEvent.Value)

  override def initializeVertex(v: Int, graph: Graph) {
    onEvent(v, graph, VertexEvent.INITIALIZE_VERTEX)
  }

  override def startVertex(v: Int, graph: Graph) {
    onEvent(v, graph, VertexEvent.START_VERTEX)
  }

  override def discoverVertex(v: Int, graph: Graph) {
    onEvent(v, graph, VertexEvent.DISCOVER_VERTEX)
  }

  override def finishVertex(v: Int, graph: Graph) {
    onEvent(v, graph, VertexEvent.FINISH_VERTEX)
  }
}